azure-pipelines.yml

```
@@ -1,813 +1,221 @@
+          parameters:
+            pythonVersion: ${{ parameters.minimumPythonVersion }}
+            testQPY: false
+            testImages: false
+
+    # The rest of the PR pipeline is to test the oldest and newest supported
+    # versions of Python, along with the integration tests (via the tutorials).
+    # It's very rare for a failure to be specific to an intermediate version of
+    # Python, so we just catch those in the cron-job pipeline to reduce the
+    # amount of resources used.
+    - stage: "Tutorials_and_Tests"
+      displayName: "Main tests"
+      dependsOn: "Lint_Docs_Prelim_Tests"
+      jobs:
+        - template: ".azure/tutorials-linux.yml"
+          parameters:
+            pythonVersion: ${{ parameters.tutorialsPythonVersion }}
+
+        - template: ".azure/test-linux.yml"
+          parameters:
+            pythonVersion: ${{ parameters.maximumPythonVersion }}
+            testQPY: true
+            testImages: true
+
+        - template: ".azure/test-macos.yml"
+          parameters:
+            pythonVersion: ${{ parameters.minimumPythonVersion }}
+
+        - template: ".azure/test-macos.yml"
+          parameters:
+            pythonVersion: ${{ parameters.maximumPythonVersion }}
+
+        - template: ".azure/test-windows.yml"
+          parameters:
+            pythonVersion: ${{ parameters.minimumPythonVersion }}
+
+        - template: ".azure/test-windows.yml"
+          parameters:
+            pythonVersion: ${{ parameters.maximumPythonVersion }}
+
+  # Push to main or the stable branches.
+  - ${{ if and(eq(variables['Build.Reason'], 'IndividualCI'), startsWith(variables['Build.SourceBranch'], 'refs/heads/')) }}:
+    - stage: "Push"
+      jobs:
+        - template: ".azure/test-linux.yml"
+          parameters:
+            pythonVersion: ${{ parameters.branchPushPythonVersion }}
+            testQPY: true
+            testImages: true
+
+  # Push to a tag.
+  - ${{ if and(eq(variables['Build.Reason'], 'IndividualCI'), startsWith(variables['Build.SourceBranch'], 'refs/tags/')) }}:
+    - stage: "Deploy"
+      jobs:
+        - template: ".azure/wheels.yml"
+          parameters:
+            jobName: "linux"
+            pool: {vmImage: 'ubuntu-latest'}
+
+        - template: ".azure/wheels.yml"
+          parameters:
+            jobName: "macos"
+            pool: {vmImage: 'macOS-10.15'}
+
+        - template: ".azure/wheels.yml"
+          parameters:
+            jobName: "macos_arm"
+            pool: {vmImage: 'macOS-10.15'}
+            env:
+              CIBW_BEFORE_ALL: rustup target add aarch64-apple-darwin
+              CIBW_ARCHS_MACOS: arm64 universal2
+              CIBW_ENVIRONMENT: CARGO_BUILD_TARGET="aarch64-apple-darwin" PYO3_CROSS_LIB_DIR="/Library/Frameworks/Python.framework/Versions/$(python -c 'import sys; print(str(sys.version_info[0])+"."+str(sys.version_info[1]))')/lib/python$(python -c 'import sys; print(str(sys.version_info[0])+"."+str(sys.version_info[1]))')"
```

```suggestion
              CIBW_ENVIRONMENT: >-
                CARGO_BUILD_TARGET="aarch64-apple-darwin"
                PYO3_CROSS_LIB_DIR="/Library/Frameworks/Python.framework/Versions/$(python -c 'import sys; print(str(sys.version_info[0])+"."+str(sys.version_info[1]))')/lib/python$(python -c 'import sys; print(str(sys.version_info[0])+"."+str(sys.version_info[1]))')"
```

Maybe a slight improvement.

---

The diff on GH for this suggestion renders in a very unhelpful way 😅